### ISSUE TITLE
[Feature Request] Implement morphological operations like dilation

### ISSUE THREAD

**Issue by ananyamukh6** [kornia/kornia#293]
https://github.com/kornia/kornia/pull/292

**Comment by edgarriba** [comment]
let's discuss here what you have in mind to implement regarding morphological operators

**Comment by ananyamukh6** [comment]
@edgarriba @ducha-aiki :
Its not differentiable, which is why I added it in a separate module different from "filters". If we plan a  `kornia.nondifferentiable` I can move it under there.
I was thinking if we had dilation and erosion, we could have a series of easily implemented morph ops like image-open, image-close, skeleton, morphological edge detection, etc.

@ducha-aiki : Do you have some reference/link/video for the "soft" or "differentiable" morphological ops (like you mentioned in the PR) that I could look into

**Comment by ducha-aiki** [comment]
@ananyamukh6 I think, this may work:
http://www.cs.tau.ac.il/~turkel/notes/Gray_Morph.pdf
Basically, it is max_pool( image window + dilation kernel) and min_pool(image_window - erosion kernel).

**Comment by edgarriba** [comment]
so what's the decision here ? I'd go for the `kornia.morphology` module.
I added in the roadmap the functionalities we already mentioned to implement.

Would a be good idea to write down here the signatures and the expected behavior/implementation since I see that this ones might be a bit tricky from traditional vs differentiable soft version.

**Comment by ananyamukh6** [comment]
@edgarriba 

for `kornia.morphology`:
Binary non-differentiable items:
1. dilation: implemented as conv2d + thresholding
- inputs must be binary, structuring element is also binary
- return (F.conv2d(img, structuring_element, padding=(se_shape[2] // 2, se_shape[2] // 2)) > 0).float()

2. erosion: 
- same input constraints as dilation
- implementation: invert image (black <-> white), perform dilation, invert image (black <-> white) (since erosion is dilation of the "background").

3. A list of derived morphological ops
- opening = erosion followed by dilation
- closing = dilation followed by erosion
- edge = image - eroded_image

Binary differentiable items:
As @ducha-aiki mentioned its a custom sliding window operation max(img + kernel), and min(img - kernel). Maybe we could do it using fold/unfold? I'd have to check

I am not sure if we are ok with non-differentiable versions or we want to have differentiable versions only. My proposal is:
- Start with non-differentiable morphological ops on a new module `kornia.morphology`.
- The brief implementation details of the non-differentiable flavors are posted above. Let me know if they look good
- My dilation branch is pretty stale (and from a fork that I have deleted), I'll open a new PR

**Comment by ananyamukh6** [comment]
Here is a first-attempt experiment on implementing dilation with fold/unfold (which might be useful for implementing grayscale/differentiable ops)

https://github.com/kornia/kornia/compare/master...ananyamukh6:ananyamukh6/dilation_with_foldunfold

The function `dilation` implements it as a conv + threshold (as I had it originally)
the function `dilation2` implements it as max(img + kernel)

`dilation2` is almost (but not quite correct). In the equation `max(img + kernel)` if the kernel is all `1`, and the image is all `0`, even then img+kernel would be `1` and max of that is `1` (whereas true dilation should have yielded an all-zero image)

In this case, too, I find that using the `max(img + kernel)` equation leaves the final image containing 1 and 2 (it should contain 0 and 1 only)

maybe the equation should be `max(img + kernel) - max(kernel)`. I used this equation in the function `dilation3` and I think I am getting the right results.

@ducha-aiki : Please take a look and see that `max(img + kernel) - max(kernel)` is the correct implementation and not `max(img + kernel)`

@edgarriba : If the new implementation looks correct, then we can decide to implement the grayscale only versions, and they will be differentiable (instead of the conv + thresholding non-differentiable one I have right now). Users can choose to input binary images and structuring elements, which would yield the classic dilation/erosion

**Comment by ducha-aiki** [comment]
@ananyamukh6 

I think, that you mixing binary dilation/erosion and grayscale. 
Results of dilation2 are indeed correct, although we might not used to it, because we think of dil/erosion as binary version only.

I will refer here to classical textbook, Digital Image Processing by Gonzalez and Woods.

![image](https://user-images.githubusercontent.com/4803565/70158219-2b546980-16b7-11ea-8ec9-22e70e3d6201.png)
![image](https://user-images.githubusercontent.com/4803565/70159190-ce59b300-16b8-11ea-970f-2f1edef13f8b.png)

![image](https://user-images.githubusercontent.com/4803565/70159536-5b9d0780-16b9-11ea-9db1-1c53eae647e1.png)

To obtain the same results with greyscale dilation, as with binary one you need to use "flat" structuring element, i.e. zeros. 
If you want to "turn-off" some elements, e.g. make "cross" structuring element, then you should use negative values, not zeros (or just -some_number).

See my notebook with examples here https://gist.github.com/ducha-aiki/83bf407451834cad7087dba09f684df7


One comment: it is only way to implement via fold/unfold? From my experience, it is much less efficient memory-wise and speed-wise than using conv/pool operations.

**Comment by ducha-aiki** [comment]
Might be relevant
https://arxiv.org/abs/2001.09046v1.pdf

**Comment by elistevens** [comment]
FYI, unless there's something obviously wrong with them, I'd suggest making exactly matching the results of the scipy.ndimage.morphology an explicit design goal. It means that users can feel confident that they can swap out one for the other, and it makes constructing test cases easier too.

Also, can't erosion be implemented with a 2d convolution of the structuring element, and then only taking the pixels that are equal to the sum of the structuring element? No need to invert, I don't think.

**Comment by ducha-aiki** [comment]
@elistevens the whole idea of kornia is to be differentiable. scipy.ndimage.morphology  is not.

**Comment by elistevens** [comment]
Ah, I thought that I'd read that `kornia.filters` was for non-differentiable things.

While it seems perhaps awful from a memory pressure perspective, could you do something like:

```python
class Erode(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 5, kernel_size=3, padding=1, bias=False)
        self.conv.weight.data.fill_(0)
        self.conv.weight.data[0,0,0,1] = 1
        self.conv.weight.data[1,0,1,0] = 1
        self.conv.weight.data[2,0,1,1] = 1
        self.conv.weight.data[3,0,1,2] = 1
        self.conv.weight.data[4,0,2,1] = 1
    def forward(self, input_mask):
        return self.conv(input_mask).min(dim=1)

class Dilate(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 5, kernel_size=3, padding=1, bias=False)
        self.conv.weight.data.fill_(0)
        self.conv.weight.data[0,0,0,1] = 1
        self.conv.weight.data[1,0,1,0] = 1
        self.conv.weight.data[2,0,1,1] = 1
        self.conv.weight.data[3,0,1,2] = 1
        self.conv.weight.data[4,0,2,1] = 1
    def forward(self, input_mask):
        return self.conv(input_mask).max(dim=1)
```

This of course assumes the standard cross structure; I think that constructing the weights wouldn't be too bad with use of `nonzero`.

**Comment by edgarriba** [comment]
in terms of API I would like to stick more to OpenCV since is what we have more or less been following. However, as we did with other modules we can take the best of the exsting frameworks and make improvements. @ducha-aiki as for differentiability we could revisit the idea of having a soft vs non-soft implementations of the functionality. I actually collected some feedback from different potential users asking for functions that maybe are not differentiable but still usable. This could  attract more users to the library.

**Comment by ducha-aiki** [comment]
@elistevens  that's actually great idea. I still think that we should do grayscale dilation and do tutorial how to get results exactly the same as for binary case, but your "terrible-memory" idea is pretty nice. 
@ananyamukh6 do you plan to continue work on this?

**Comment by edgarriba** [comment]
adding this to the conversation: https://www.kaggle.com/ren4yu/bengali-morphological-ops-as-image-augmentation

we discussed with @ducha-aiki to use the challenge as a use case to define the API. Maybe could start by mimicking OpenCV and later iterate on that.

**Comment by ducha-aiki** [comment]
OK, here is a minimalistic implementation of proper grayscale morphology based on @elistevens idea

```python3
def se_to_mask(se: torch.Tensor) -> torch.Tensor:
    se_h, se_w = se.size()
    se_flat = se.view(-1)
    num_feats = se_h * se_w
    out = torch.zeros(num_feats, 1, se_h, se_w, dtype = se.dtype, device=se.device)
    for i in range(num_feats):
        y = i % se_h
        x = i // se_h
        out[i,0, x, y] = (se_flat[i] >= 0).float() ;
    return  out

class Dilate(nn.Module):
    def __init__(self, se: torch.tensor):
        super().__init__()
        self.se = se
        self.se_h = se.size(0)
        self.se_w = se.size(1)
    def forward(self, input):
        return (F.conv2d(input, se_to_mask(self.se)) + self.se.view(1,-1,1,1)).max(dim=1)[0]
    

class Erode(nn.Module):
    def __init__(self, se: torch.tensor):
        super().__init__()
        self.se = se
        self.se_h = se.size(0)
        self.se_w = se.size(1)
    def forward(self, input):
        return (F.conv2d(input, se_to_mask(self.se)) - self.se.view(1,-1,1,1)).min(dim=1)[0]
```

To have results exactly as in binary dilation, one should use (-1, 0) kernels, where -1 for "off" elements and "0" for on. 
e. g. cross will be:

```python3
kernel_cross = torch.tensor([[0,1,0],[1,1,1], [0,1,0]]) - 1.
```

The implementation above lacks fast se_to_mask implementation, i.e. without for loop and also would benefit from adding strides, padding and, of course, documentation.

Would anyone like to finish this and do PR? 
@ananyamukh6 @elistevens

**Comment by elistevens** [comment]
Unfortunately I'm under deadline elsewhere, and that's soaking up the majority of my free time ATM. I don't think I'd be able to take something to completion until late Feb. or March. I don't know what schedule you're thinking of.

A thought:  `se_to_mask` could be called in `__init__` rather than in `forward` to cache the result.

**Comment by edgarriba** [comment]
cool, as for the function signature I would stick to opencv:

#### Dilate
https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#ga4ff0f3318642c4f469d0e11f242f3b6c

```python

def dilate(input: torch.Tensor, kernel: torch.Tensor, anchor: torch.Tensor, border_type: str = 'reflect') -> torch.Tensor:
    return Dilate(...)(input)
```

#### Erode
https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#gaeb1e0c1033e3f6b891a25d0511362aeb

```python

def erode(input: torch.Tensor, kernel: torch.Tensor, anchor: torch.Tensor, border_type: str = 'reflect') -> torch.Tensor:
    return Erode(...)(input)
```
and also adding a kernel generation function:

#### GetStructuringElement
https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#gac342a1bb6eabf6f55c803b09268e36dc

https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#gac2db39b56866583a95a5680313c314ad

```python
def get_structuring_element(shape: str, kernel_size: int, anchor: Optional[torch.Tensor] = None) -> torch.Tensor:
        assert shape in ('rectangular', 'cross', 'ellipse',)
        # create the_kernel_tensor
    return (the_kernel_tensor)
```

**Comment by ducha-aiki** [comment]
Do we really need anchor point? It compicates things a lot. And do people really use filters/SEs, which are not centered? I never even heard about this.  Moreover, we already ignored it in our filter2d)) 
https://docs.opencv.org/master/d4/d86/group__imgproc__filter.html#ga27c049795ce870216ddfb366086b5a04

**Comment by edgarriba** [comment]
fair enough, lets ignore it

**Comment by edgarriba** [comment]
There's an offline implementation process going on this.

**Comment by edgarriba** [comment]
merged in https://github.com/kornia/kornia/pull/831
